tip-distributor/src/claim_mev_workflow.rs

```diff
@@ -32,121 +41,426 @@ pub enum ClaimMevError {
     JsonError(#[from] serde_json::Error),
 }
 
-pub fn claim_mev_tips(
+pub async fn claim_mev_tips(
     merkle_root_path: &PathBuf,
-    rpc_url: &str,
+    rpc_url: String,
+    rpc_connection_count: u64,
+    max_concurrent_rpc_get_reqs: usize,
     tip_distribution_program_id: &Pubkey,
     keypair_path: &PathBuf,
+    max_loop_retries: u64,
+    max_loop_duration: Duration,
 ) -> Result<(), ClaimMevError> {
-    const MAX_RETRY_DURATION: Duration = Duration::from_secs(600);
-
     let merkle_trees: GeneratedMerkleTreeCollection =
         read_json_from_file(merkle_root_path).expect("read GeneratedMerkleTreeCollection");
-    let keypair = read_keypair_file(keypair_path).expect("read keypair file");
+    let keypair = Arc::new(read_keypair_file(keypair_path).expect("read keypair file"));
+    let payer_pubkey = keypair.pubkey();
+    let blockhash_rpc_client = Arc::new(RpcClient::new_with_commitment(
+        rpc_url.clone(),
+        CommitmentConfig::finalized(),
+    ));
+    let rpc_clients = Arc::new(
+        (0..rpc_connection_count)
+            .map(|_| {
+                Arc::new(RpcClient::new_with_commitment(
+                    rpc_url.clone(),
+                    CommitmentConfig::confirmed(),
+                ))
+            })
+            .collect_vec(),
+    );
 
-    let tip_distribution_config =
-        Pubkey::find_program_address(&[Config::SEED], tip_distribution_program_id).0;
+    let tree_nodes = merkle_trees
+        .generated_merkle_trees
+        .iter()
+        .flat_map(|tree| &tree.tree_nodes)
+        .collect_vec();
+    let stake_acct_min_rent = blockhash_rpc_client
+        .get_minimum_balance_for_rent_exemption(StakeState::size_of())
+        .await
+        .expect("Failed to calculate min rent");
 
-    let rpc_client =
-        RpcClient::new_with_commitment(rpc_url.to_string(), CommitmentConfig::finalized());
+    // fetch all accounts up front
+    info!("Starting to fetch accounts");
+    let account_fetch_start = Instant::now();
+    let tdas = get_batched_accounts(
+        &blockhash_rpc_client,
+        max_concurrent_rpc_get_reqs,
+        merkle_trees
+            .generated_merkle_trees
+            .iter()
+            .map(|tree| tree.tip_distribution_account)
+            .collect_vec(),
+    )
+    .await
+    .unwrap()
+    .into_iter()
+    .filter_map(|(pubkey, maybe_account)| {
+        let account = match maybe_account {
+            Some(account) => account,
+            None => {
```

Comment:
you panic below if it's not in the hashmap, why not just early return error here; and have a `Vec<(GenMerkTree, Tda)>`

Reply:
a bit difficult because we fetch the accounts in batches

Reply:
return errors now